watertap/core/plugins/solvers.py

```diff
@@ -0,0 +1,150 @@
+class IpoptWaterTAP(IPOPT):
+
+    def __init__(self, **kwds):
+        kwds["name"] = "ipopt-watertap"
+        super().__init__(**kwds)
+
+    def _presolve(self, *args, **kwds):
+        if len(args) > 1 or len(args) == 0:
+            raise TypeError(f"IpoptWaterTAP.solve takes 1 positional argument but {len(args)} were given")
+        if not isinstance(args[0], (_BlockData, IBlock)):
+            raise TypeError("IpoptWaterTAP.solve takes 1 positional argument: a Pyomo ConcreteModel or Block")
+
+        self._tee = kwds.get("tee", False)
+
+        # Set the IDAES-default tol
+        if "tol" not in self.options:
+            self.options["tol"] = 1e-6
+
+        if not self._is_user_scaling():
+            self._reset_needed = False
+            super()._presolve(*args, **kwds)
+            return
+
+        if self._tee:
+            print("ipopt-watertap: Ipopt with user variable scaling and IDAES jacobian constraint scaling")
+
+        # These options are typically available with grandient-scaling, and they
+        # have corresponding options in the IDAES constraint_autoscale_large_jac
+        # function. Here we use there Ipopt names and default values, see
+        # https://coin-or.github.io/Ipopt/OPTIONS.html#OPT_NLP_Scaling
+        max_grad = self._get_option("nlp_scaling_max_gradient", 100)
+        min_scale = self._get_option("nlp_scaling_min_value", 1e-8)
+
+        # These options are custom for the IDAES constraint_autoscale_large_jac
+        # function. We expose them as solver options as this has become part
+        # of the solve process.
+        ignore_variable_scaling = self._get_option("ignore_variable_scaling", False)
+        ignore_constraint_scaling = self._get_option("ignore_constraint_scaling", False)
+
+        self._model = args[0]
+        self._cache_scaling_factors()
+        self._reset_needed = True
+
+        # NOTE: This function sets the scaling factors on the
+        #       constraints. Hence we cache the constraint scaling
+        #       factors and reset them to their original values
+        #       so that repeated calls to solve change the scaling
+        #       each time based on the initial values, just like in Ipopt.
+        try:
+            constraint_autoscale_large_jac(self._model,
+                    ignore_constraint_scaling=ignore_constraint_scaling,
+                    ignore_variable_scaling=ignore_variable_scaling,
+                    max_grad=max_grad,
+                    min_scale=min_scale)
+        except AssertionError as err:
+            if str(err) == "Error in AMPL evaluation":
+                print("ipopt-watertap: Issue in AMPL function evaluation; Jacobian constraint scaling not applied.")
+                halt_on_ampl_error = self.options.get("halt_on_ampl_error", "yes")
+                if halt_on_ampl_error == "no" :
+                    print("ipopt-watertap: halt_on_ampl_error=no, so continuing with optimization.")
+                else:
```

Comment:
This question is just for my understanding because I'm not sure about how the logic works here. What gets assigned to `halt_on_ampl_error` from `self.options.get("halt_on_ampl_error", "yes")`? If halt_on_ampl_error was set to no, what would be assigned to `halt_on_ampl_error`? 

Also, in the `else` clause, the error message tells the user to run with `halt_on_ampl_error` set to yes. Isn't it already set to yes? Or is the purpose of the error message just to tell the user to _also_ set `symbolic_solver_labels=True` to more clearly diagnose the issue?

Reply:
`halt_on_ampl_error = self.options.get("halt_on_ampl_error", "yes")` is equivalent to
```
if "halt_on_ampl_error" not in self.options:
    halt_on_ampl_error = "yes"
else:
    halt_on_ampl_error = self.options["halt_on_ampl_error"]
```

Reply:
Regarding the second point -- the default Ipopt option is `halt_on_ampl_error=no`. But given what we're doing, I think it makes sense to have a more cautions default for this solver, and if you run Ipopt with these two options it will print out the (likely) problematic constraint in a way you can use to diagnose your Pyomo model.